### PR TITLE
+ add revenue distributor and buyback

### PR DESCRIPTION
**Pull request by tedw-lista**

### CODE REVIEW

File: contracts/dao/ListaAutoBuyBack.sol [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        require(success, "1inch call failed");
+        (uint256 amountOut,) = abi.decode(result, (uint256, uint256));
+
+        uint256 today = block.timestamp / DAY * DAY;
+        dailyBought[today] = dailyBought[today] + amountOut;
+
+        emit BoughtBack(_tokenIn, _amountIn, amountOut);
+    }
+
+    function managerTransfer(address _token, uint256 _amount) external onlyRole(MANAGER) {
```
Comment (eason-id) [inline contracts/dao/ListaAutoBuyBack.sol:93]: withdraw to the` msg.sender`

File: contracts/dao/ListaAutoBuyBack.sol [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+import "@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol";
+import "@openzeppelin/contracts/utils/math/Math.sol";
+
+/**
+  * @title ListaAutoBuyBack
+  * @dev lista revenue auto buy back
+  * @dev all tokens balance of ListaAutoBuyBack will be swapped to Lista Token using 1inch router
+  * @dev result of swap will be sent to receiver address to distribute to users
+  */
+contract ListaAutoBuyBack is Initializable, AccessControlUpgradeable {
```
Comment (eason-id) [inline contracts/dao/ListaAutoBuyBack.sol:16]: buyback is a wold

Comment (qingyang-lista) [inline contracts/dao/ListaAutoBuyBack.sol:16]: buyback is a wold

File: contracts/dao/ListaAutoBuyBack.sol [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        (uint256 amountOut,) = abi.decode(result, (uint256, uint256));
+
+        uint256 today = block.timestamp / DAY * DAY;
+        dailyBought[today] = dailyBought[today] + amountOut;
+
+        emit BoughtBack(_tokenIn, _amountIn, amountOut);
+    }
+
+    function managerTransfer(address _token, uint256 _amount) external onlyRole(MANAGER) {
+        IERC20(_token).safeTransfer(defaultReceiver, _amount);
```
Comment (qingyang-lista) [inline contracts/dao/ListaAutoBuyBack.sol:94]: withdraw to `msg.sender`

File: contracts/dao/ListaAutoBuyback.sol [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        }
+
+        (uint256 amountOut,) = abi.decode(result, (uint256, uint256));
+        uint256 today = block.timestamp / DAY * DAY;
+        dailyBought[today] = dailyBought[today] + amountOut;
+
+        emit BoughtBack(_tokenIn, _amountIn, amountOut);
+    }
+
+    function managerTransfer(address _token, uint256 _amount) external onlyRole(MANAGER) {
```
Comment (qingyang-lista) [inline contracts/dao/ListaAutoBuyback.sol:106]: The role should be `admin`, not` bot`

File: contracts/dao/ListaRevenueDistributor.sol [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        revenueWalletAddress = _revenueWalletAddress;
+        listaDistributeToAddress = _listaDistributeToAddress;
+        distributeRate = _distributeRate;
+    }
+
+    /**
+     * @dev distribute tokens to autoBuybackAddress and revenueWalletAddress according to distributeRate
+     * @param tokens, tokens to distribute
+     */
+    function distributeTokens(address[] memory tokens) external onlyRole(MANAGER) {
```
Comment (qingyang-lista) [inline contracts/dao/ListaRevenueDistributor.sol:80]: Add an token whitelist in contract for checking  whether the tokens exist in token whitelist

**Review by qingyang-lista — APPROVED** [review]
LGTM

**Pull request merged**